### ISSUE TITLE
Can you list required versions? Have MMCV 2.1.0 installed but mmyolo says it needs mmcv 2.0< so it gets uninstalled / upgraded on launch

### ISSUE THREAD

**Issue by CCpt5** [wkpark/uddetailer#54]
Hello again! On a different extension of yours this time! :)

Thanks for taking this on - I never actually used the old one, but seeing that different checkpoints can be used for the after generation touchups, I'm excited that it may be possible to fix the "similar faces" or have two different people added to images.  

But my problem is install. I've had a very difficult time. I read your note about the thread for issues w/ the mmcv wheel/install. I got past that after a bunch of trial and error. However, now I'm stuck w/ an incompatibility between mmcv and mmyolo. The latest mmcv is 2.1.0 and I was on that, but downgraded to 2.0.0 which failed, and even tried 1.7.0 full per this  thread about the same error I was getter from the original repo: https://github.com/dustysys/ddetailer/issues/38#issuecomment-1507838400  
  
The solution there seems like it'll work which is: 

```
Found a solution to 'No module named 'mmdet.core''. Maybe because mmdet==3.0.0 has no core.
I uninstall mmdet 3.0.0 then install mmdet 2.25.3 and the problem is fixed.
Note that mmdet 2.25.3 is incompatible with mmcv>=2.0.0, so you also need to uninstall mmcv 2.0.0 and install mmcv-full 1.7.0.
Hopes this will help.
```

But again when I launch A1111 it checks the versions and automatically updates mmcv. 

Here is my launch console. If you can advise on proper versions to use that will not trigged changes on launch, or if necessary a way to disable checking/changing versions I'd very much appreciate it!

```
venv "D:\Stable-Diffusion-Webui\venv\Scripts\Python.exe"
Python 3.10.11 (tags/v3.10.11:7d4cc5a, Apr  5 2023, 00:38:17) [MSC v.1929 64 bit (AMD64)]
Version: v1.6.0-2-g4afaaf8a
Commit hash: 4afaaf8a020c1df457bcf7250cb1c7f609699fa7
Installing frame2frame requirement: opencv-python
Found existing installation: mmdet 2.28.0
Uninstalling mmdet-2.28.0:
  Successfully uninstalled mmdet-2.28.0
Found existing installation: mmcv-full 1.7.0
Uninstalling mmcv-full-1.7.0:
  Successfully uninstalled mmcv-full-1.7.0
Uninstalling mmdet, mmcv, mmcv-full... (if installed)
Installing mmcv, mmdet, mmengine...
Installing mmcv
Installing mmdet
Installing mmdet
Check mmcv version...
Your mmcv version 2.1.0 may not work mmyolo.
Please install mmcv version 2.0.0 manually or uninstall mmcv and restart UI again to install mmcv 2.0.0
Check mmengine version...
your mmengine version is 0.8.5
Launching Web UI with arguments: --opt-sdp-attention --no-half-vae --opt-channelslast --disable-safe-unpickle --skip-torch-cuda-test --disable-nan-check --skip-version-check --ckpt-dir e:\stable Diffusion Checkpoints
no module 'xformers'. Processing without...
no module 'xformers'. Processing without...
No module 'xformers'. Proceeding without it.
[-] ADetailer initialized. version: 23.11.1, num models: 9
2023-11-27 20:30:34,011 - ControlNet - INFO - ControlNet v1.1.419
ControlNet preprocessor location: D:\Stable-Diffusion-Webui\extensions\3sd-webui-controlnet\annotator\downloads
2023-11-27 20:30:34,062 - ControlNet - INFO - ControlNet v1.1.419
Tag Autocomplete: Could not locate model-keyword extension, Lora trigger word completion will be limited to those added through the extra networks menu.
[sd-webui-freeu] Controlnet support: *enabled*
ERROR - failed to import mmyolo - MMCV==2.1.0 is used but incompatible. Please install mmcv>=2.0.0rc4, <2.1.0.
Check config files...
Done
 Total 4 mmdet, 0 yolo and 3 mediapipe models.
SUCCESS - success to load config for D:\Stable-Diffusion-Webui\models\mmdet\segm\mmdet_dd-person_mask2former.pth!
 You can enable model validity tester in the Settings-> μ DDetailer.
SUCCESS - success to load config for D:\Stable-Diffusion-Webui\models\mmdet\bbox\mmdet_anime-face_yolov3.pth!
SUCCESS - success to load config for D:\Stable-Diffusion-Webui\models\mmdet\segm\yolov5_ins_n.pth!
SUCCESS - success to load config for D:\Stable-Diffusion-Webui\models\mmdet\segm\yolov5_ins_s.pth!
 Total 4 valid mmdet configs are found.
 You can disable validity tester in the Settings-> μ DDetailer.
Loading weights [74dda471cc] from e:\stable Diffusion Checkpoints\SDXL\realvisxlV20_v20Bakedvae.safetensors
2023-11-27 20:30:35,375 - AnimateDiff - INFO - Injecting LCM to UI.
Creating model from config: D:\Stable-Diffusion-Webui\repositories\generative-models\configs\inference\sd_xl_base.yaml
Loading VAE weights specified in settings: D:\Stable-Diffusion-Webui\models\VAE\sdxl_vae.safetensors
Applying attention optimization: sdp-no-mem... done.
Model loaded in 3.6s (load weights from disk: 0.6s, create model: 0.2s, apply weights to model: 2.2s, load VAE: 0.1s, load textual inversion embeddings: 0.1s, calculate empty prompt: 0.1s).
loading.. cn_modole
 - ControlNet extension 3sd-webui-controlnet found
 - default µ DDetailer model= bbox/mmdet_anime-face_yolov3.pth [38208bb6]
2023-11-27 20:30:39,066 - AnimateDiff - INFO - Hacking i2i-batch.
 - default µ DDetailer model= bbox/mmdet_anime-face_yolov3.pth [38208bb6]
Running on local URL:  http://127.0.0.1:7860

To create a public link, set `share=True` in `launch()`.
Startup time: 20.6s (prepare environment: 10.3s, import torch: 1.9s, import gradio: 0.4s, setup paths: 0.3s, initialize shared: 0.2s, other imports: 0.4s, list SD models: 0.2s, load scripts: 2.2s, create ui: 4.2s, gradio launch: 0.4s, app_started_callback: 0.1s).
```

I also used Chatgpt4 premium to try get this working and if interested here are shareable links to those chats - it shows where I started (different errors gettign mmcv install which I looked up and fixed per your front page) but then others until ultimately where I'm at where I think all versions will work but the laucnh check is changing them back to incompatible versions.

Chatgpt logs from today:

https://chat.openai.com/share/e86cfded-6cf0-4467-b34b-debbc1ab59bf

**Comment by wkpark** [comment]
thank you for your reporting in another extension :)

this is a short note.

currently Im using `mmcv == 2.1.0`, `mmdet >=3.0.0`, `mmengine == 0.9.1` under windows, with mmyolo `0.6.0`

- patched mmyolo 0.6.0. - simply fix version restriction at `venv/Lib/site-packages/mmyolo/__init__.py` (fix  `mmcv_maximum_version = '2.1.0'` to `mmcv_maximum_version = '2.1.1'` works fine.
- mmengine 0.9.1 + patched `bitesnbytes` at https://github.com/wkpark/bitsandbytes/actions/runs/6887358987 (login github and scroll down you will find an downloadable artifacts)

**Comment by wkpark** [comment]
ahh. my mistake. if so,.... downgrade `mmengine` `0.8.5` will work.

**Comment by wkpark** [comment]
> Ok, so I tried 0.8.5 and then got an error about a missing controlnet file(cn_module.py?). Buuuut after getting that error, I decided to delete the extension folder, and this time Git clone it directly from this repo. I had first seen uddetailer on the A1111 extension list a few days ago and installed it using the in-app installer. I'm not sure if using the app installer could have issues properly installing/setting this up, but after deleting the folder and installing directly I no longer have any uddetailer errors on launch!
> 
> If you find other users having issues w/ installing the extension you may want to try installing it yourself from the in-app installer to make sure that it works on your end. I'm happy it seems to be setup properly now so I'm not going to attempt an uninstall/reinstall from the app right now :)
> 
> BTW if you have a kofi or anything like that drop a link sometime - appreciate your efforts resurrecting these extensions giving them new life w/ all the new advancements the past months......

Thank you for your patience and feedback!

Lots of people install and use it, but not many are willing to give any feedback or fix it - thanks for that.

and this is my ko-fi https://ko-fi.com/wkpark :)

**Comment by rafstahelin** [comment]
> thank you for your reporting in another extension :)
> 
> this is a short note.
> 
> currently Im using `mmcv == 2.1.0`, `mmdet >=3.0.0`, `mmengine == 0.9.1` under windows, with mmyolo `0.6.0`
> 
> * patched mmyolo 0.6.0. - simply fix version restriction at `venv/Lib/site-packages/mmyolo/__init__.py` (fix  `mmcv_maximum_version = '2.1.0'` to `mmcv_maximum_version = '2.1.1'` works fine.
> * mmengine 0.9.1 + patched `bitsnbytes` at https://github.com/wkpark/bitsandbytes/actions/runs/6887358987 (login github and scroll down you will find a downloadable artifacts)

hello
is still worth fixing? 

Currently:
Your mmcv version 2.1.0 may not work mmyolo.
*** Please install mmcv version 2.0.0 manually or uninstall mmcv and restart UI again to install mmcv 2.0.0
*** Check mmengine version...
*** your mmengine version is 0.8.5
*** Installing mmyolo

Should i downgrade to 2.0.0

I tried git cloning instead of in-app, still broken. Please do let me know. In Windows 11. Python 3.10.9 Cuda 12.1

**Comment by wkpark** [comment]
> > thank you for your reporting in another extension :)
> > this is a short note.
> > currently Im using `mmcv == 2.1.0`, `mmdet >=3.0.0`, `mmengine == 0.9.1` under windows, with mmyolo `0.6.0`
> > 
> > * patched mmyolo 0.6.0. - simply fix version restriction at `venv/Lib/site-packages/mmyolo/__init__.py` (fix  `mmcv_maximum_version = '2.1.0'` to `mmcv_maximum_version = '2.1.1'` works fine.
> > * mmengine 0.9.1 + patched `bitsnbytes` at https://github.com/wkpark/bitsandbytes/actions/runs/6887358987 (login github and scroll down you will find a downloadable artifacts)
> 
> hello is still worth fixing?
>

Yes. mmcv 2.1.1 works fine with manual fix.

> Currently: Your mmcv version 2.1.0 may not work mmyolo. *** Please install mmcv version 2.0.0 manually or uninstall mmcv and restart UI again to install mmcv 2.0.0 *** Check mmengine version... *** your mmengine version is 0.8.5 *** Installing mmyolo
> 
> Should i downgrade to 2.0.0
> 
> I tried git cloning instead of in-app, still broken. Please do let me know. In Windows 11. Python 3.10.9 Cuda 12.1

Please see also https://github.com/wkpark/uddetailer/wiki/Required-python-packages

**Comment by rafstahelin** [comment]
i dont see mmyolo at "venv/Lib/site-packages/mmyolo/__init__.py" 
Also, noticing mmengine 0.9.1 only seems to be available for Cuda 11.8? I am running in 12.1 in python 3.10.9
I installed via git clone in my windows 11
<img width="233" alt="image" src="https://github.com/wkpark/uddetailer/assets/64715158/acf776b7-616b-4824-8f30-d7540e5a7795">
please advise thanks

**Comment by rafstahelin** [comment]
I should add, I am in Forge webi

**Comment by wkpark** [comment]
> I should add, I am in Forge webi

without mmdet/mmcv/mmyolo, you can use mediapipe models or ultralytics yolov8 models

I can confirm that `mmcv==2.1.0` + `mmengine==0.8.5` + `mmyolo 0.6` (with fix)
~~~bash
(venv) D:\src\stable-diffusion-webui-forge>pip show mmyolo
Name: mmyolo
Version: 0.6.0
Summary: OpenMMLab Toolbox of YOLO
Home-page: https://github.com/open-mmlab/mmyolo
Author: MMYOLO Contributors
Author-email: openmmlab@gmail.com
License: GPL License 3.0
Location: d:\src\stable-diffusion-webui-forge\venv\lib\site-packages
Requires: numpy, prettytable
Required-by:

(venv) D:\src\stable-diffusion-webui-forge>pip show mmengine
Name: mmengine
Version: 0.8.5
Summary: Engine of OpenMMLab projects
Home-page: https://github.com/open-mmlab/mmengine
Author: MMEngine Authors
Author-email: openmmlab@gmail.com
License: UNKNOWN
Location: d:\src\stable-diffusion-webui-forge\venv\lib\site-packages
Requires: addict, matplotlib, numpy, opencv-python, pyyaml, regex, rich, termcolor, yapf
Required-by: mmcv

(venv) D:\src\stable-diffusion-webui-forge>pip show mmcv
Name: mmcv
Version: 2.1.0
Summary: OpenMMLab Computer Vision Foundation
Home-page: https://github.com/open-mmlab/mmcv
Author: MMCV Contributors
Author-email: openmmlab@gmail.com
License:
Location: d:\src\stable-diffusion-webui-forge\venv\lib\site-packages
Requires: addict, mmengine, numpy, opencv-python, packaging, Pillow, pyyaml, regex, yapf
Required-by:
~~~

**Comment by rafstahelin** [comment]
How would I do that. I'm getting the massive mmcv error nesssage. How do I replace mmcv fir other model? Do you have a step by step please?

**Comment by wkpark** [comment]
> How would I do that. I'm getting the massive mmcv error nesssage. How do I replace mmcv fir other model? Do you have a step by step please?

open cmd prompt and install mmcv again. for example
~~~bash
cd to_your_sdwebui-forge # goto sdwebu-forge
venv\scripts\activate # activate venv
pip show mmcv # show current mmcv version
~~~

and install mmcv again using `mim`
~~~
mim install mmcv==2.1.0
~~~

**Comment by rafstahelin** [comment]
still dont see mmyolo folder with init. what else do i need to do after installing 2.1.0?

**Comment by wkpark** [comment]
> still dont see mmyolo folder with init. what else do i need to do after installing 2.1.0?

you can install `mmyolo` using `mim install` command.
 - `mim install mmyolo` (automatically installed by installer)
 - or `python -m mim install mmyolo`

(there seems to be an issue as `mmyolo` is not installing automatically, I guess)

all other python packages could be installed using `pip install`. for example
 - `pip install ultralytics` (you need to install it manually)

**Comment by rafstahelin** [comment]
ok, when i run mim install mmyolo i get: 
 RuntimeError:
`      The detected CUDA version (11.8) mismatches the version that was used to compile
      PyTorch (12.1). Please make sure to use the same CUDA versions.`

I compiled Forge webui using 12.1 cuda

`note: This error originates from a subprocess, and is likely not a problem with pip.
  Rolling back uninstall of mmcv
  Moving to e:\forge-webui\venv\lib\site-packages\mmcv-2.1.0.dist-info\
   from E:\forge-webui\venv\Lib\site-packages\~mcv-2.1.0.dist-info
  Moving to e:\forge-webui\venv\lib\site-packages\mmcv\
   from E:\forge-webui\venv\Lib\site-packages\~mcv
error: legacy-install-failure

× Encountered error while trying to install package.
╰─> mmcv

note: This is an issue with the package mentioned above, not pip.
hint: See above for output from the failure.`

Does it not work with 12.1?

**Comment by wkpark** [comment]
recommended mmengine version is `mmengine-0.8.5`.
my torch version is 2.1.2+cu118, so I can't test/tell your settings is working or not, sorry.

but, there are official document exist for mmcv+cu121:
https://mmcv.readthedocs.io/en/latest/get_started/installation.html#install-mmcv
and it says, to install `mmcv+cu121`
`pip install mmcv==2.1.0 -f https://download.openmmlab.com/mmcv/dist/cu121/torch2.1/index.html`

**Comment by rafstahelin** [comment]
yea alright, not gonna work for people on cuda 12.1
thanks for your help

**Comment by rafstahelin** [comment]
ok, so i went ahead and installed Pytorch2.1.2+Cuda11.8 like you and ran with xformers.
mmcv 2.1.0 doesnt work for me. I guess I will just run as you say with mediapipe and yolo9
thanks

Here's the console:
`venv "E:\stable-diffusion-webui-forge\venv\Scripts\Python.exe"
Python 3.10.9 (tags/v3.10.9:1dd9be6, Dec  6 2022, 20:01:21) [MSC v.1934 64 bit (AMD64)]
Version: f0.0.12-latest-155-gd81e353d
Commit hash: d81e353d8928147bbd973068d0efbb2802affe0f
*** Error running install.py for extension E:\stable-diffusion-webui-forge\extensions\uddetailer.
*** Command: "E:\stable-diffusion-webui-forge\venv\Scripts\python.exe" "E:\stable-diffusion-webui-forge\extensions\uddetailer\install.py"
*** Error code: 1
*** stdout: Check mmcv version...
*** Your mmcv version 2.1.0 may not work mmyolo.
*** Please install mmcv version 2.0.0 manually or uninstall mmcv and restart UI again to install mmcv 2.0.0
*** Check mmengine version...
*** your mmengine version is 0.8.5
*** Installing mmyolo
***
*** stderr: Traceback (most recent call last):
***   File "E:\stable-diffusion-webui-forge\extensions\uddetailer\install.py", line 123, in <module>
***     install()
***   File "E:\stable-diffusion-webui-forge\extensions\uddetailer\install.py", line 95, in install
***     run(f'"{python}" -m mim install mmyolo', desc="Installing mmyolo", errdesc="Couldn't install mmyolo")
***   File "E:\stable-diffusion-webui-forge\modules\launch_utils.py", line 124, in run
***     raise RuntimeError("\n".join(error_bits))
*** RuntimeError: Couldn't install mmyolo.
*** Command: "E:\stable-diffusion-webui-forge\venv\Scripts\python.exe" -m mim install mmyolo
*** Error code: 1
*** stdout: Looking in links: https://download.openmmlab.com/mmcv/dist/cu118/torch2.1.0/index.html
*** Collecting mmyolo
***   Using cached mmyolo-0.6.0-py3-none-any.whl.metadata (23 kB)
*** Requirement already satisfied: numpy in e:\stable-diffusion-webui-forge\venv\lib\site-packages (from mmyolo) (1.26.2)
*** Requirement already satisfied: prettytable in e:\stable-diffusion-webui-forge\venv\lib\site-packages (from mmyolo) (3.9.0)
*** Collecting mmcv<2.1.0,>=2.0.0rc4 (from mmyolo)
***   Using cached mmcv-2.0.1.tar.gz (469 kB)
***   Preparing metadata (setup.py): started
***   Preparing metadata (setup.py): finished with status 'done'`

**Comment by wkpark** [comment]
It seems `mim install mmyolo` command tried to install `mmcv==2.0.1` again with cuda 11.8 build. and it cause the issue.

I suggest you the following command:
- `pip install mmcv==2.1.0 -f https://download.openmmlab.com/mmcv/dist/cu121/torch2.1/index.html`  # install mmcv first.
- `pip install mmyolo`

**Comment by rafstahelin** [comment]
Ok will try Will also try the 12.1 suggestion from yesterday Best, Raf On 16 Feb 2024, at 02:44, Won-Kyu Park ***@***.***> wrote:﻿
It seems mim install mmyolo command tried to install mmcv==2.0.1 again with cuda 11.8 build. and it cause the issue.
I suggest you the following command:

pip install mmcv==2.1.0 -f https://download.openmmlab.com/mmcv/dist/cu121/torch2.1/index.html  # install mmcv first.
pip install mmyolo

**Comment by rafstahelin** [comment]
tried the above on my 12.1 torch install (i tried different forge webui installations with 11.8 and 12.1)
With the 12.1 torch install, and your pop install 2.1.0  then mmyolo i got a clean console, but when i launched webui i go this:
`ERROR - failed to import mmyolo - MMCV==2.1.0 is used but incompatible. Please install mmcv>=2.0.0rc4, <2.1.0.`
I guess i am gonna run without mmyolo. 

How do i remove this from the startup? I dont want this message on every startup. Thanks

**Comment by wkpark** [comment]
> tried the above on my 12.1 torch install (i tried different forge webui installations with 11.8 and 12.1) With the 12.1 torch install, and your pop install 2.1.0 then mmyolo i got a clean console, but when i launched webui i go this: `ERROR - failed to import mmyolo - MMCV==2.1.0 is used but incompatible. Please install mmcv>=2.0.0rc4, <2.1.0.` I guess i am gonna run without mmyolo.
> 
> How do i remove this from the startup? I dont want this message on every startup. Thanks

no. no. that's good message.
after install mmyolo, you have to fix `__init__.py` of mmyolo file. that's was missing point.

**Comment by rafstahelin** [comment]
> > tried the above on my 12.1 torch install (i tried different forge webui installations with 11.8 and 12.1) With the 12.1 torch install, and your pop install 2.1.0 then mmyolo i got a clean console, but when i launched webui i go this: `ERROR - failed to import mmyolo - MMCV==2.1.0 is used but incompatible. Please install mmcv>=2.0.0rc4, <2.1.0.` I guess i am gonna run without mmyolo.
> > How do i remove this from the startup? I dont want this message on every startup. Thanks
> 
> no. no. that's good message. after install mmyolo, you have to fix `__init__.py` of mmyolo file. that's was missing point.
> 
> ```
> patched mmyolo 0.6.0. - simply fix version restriction at venv/Lib/site-packages/mmyolo/__init__.py (fix mmcv_maximum_version = '2.1.0' to mmcv_maximum_version = '2.1.1' works fine.
> ```
> 
> ## TL;DR
> * `torch 2.0.x` - `mmcv 2.0.1` + `mmyolo 0.6.0` - OK
> * `torch 2.1.x` - `mmcv 2.1.1` + `mmyolo 0.6.0` - need to fix version restriction of `__init__.py` of `mmyolo` manually.

gotcha. the problem is i dont see a mmyolo folder in site-packages:
<img width="375" alt="image" src="https://github.com/wkpark/uddetailer/assets/64715158/7ba0a23d-5026-4770-a2dd-65239defa584">

**Comment by wkpark** [comment]
`ERROR - failed to import mmyolo - MMCV==2.1.0 is used but incompatible. Please install mmcv>=2.0.0rc4, <2.1.0.`
this message is from mmyolo package. I guess, you've install `mmyolo` without `venv` environment.

in this case, `deactivate` venv first, and `pip uninstall mmyolo`, then `activate` venv again and `pip install mmyolo`

**Comment by rafstahelin** [comment]
E:\forge-webui>pip uninstall mmyolo
WARNING: Skipping mmyolo as it is not installed.

Weird. It is not installed

**Comment by rafstahelin** [comment]
success!!

`
<img width="722" alt="image" src="https://github.com/wkpark/uddetailer/assets/64715158/654a79de-8363-4d88-83f3-629dcf4213dd">
`

**Comment by rafstahelin** [comment]
I just uninstalled it inside my venv the reinstalled it mmyolo and changed to amx 2.1.1 in the init
Dont know why it didnt work the first time around
thanks for your help

**Comment by wkpark** [comment]
fixed controlnet support for forge. PR #75

**Comment by galavera** [comment]
> thank you for your reporting in another extension :)
> 
> this is a short note.
> 
> currently Im using `mmcv == 2.1.0`, `mmdet >=3.0.0`, `mmengine == 0.9.1` under windows, with mmyolo `0.6.0`
> 
> * patched mmyolo 0.6.0. - simply fix version restriction at `venv/Lib/site-packages/mmyolo/__init__.py` (fix  `mmcv_maximum_version = '2.1.0'` to `mmcv_maximum_version = '2.1.1'` works fine.
> * mmengine 0.9.1 + patched `bitsnbytes` at https://github.com/wkpark/bitsandbytes/actions/runs/6887358987 (login github and scroll down you will find a downloadable artifacts)

I uninstalled mmyolo, reinstalled, changed __init__ max version settings, and I'm still getting this error. removed pycache, restarted system as well after making changes.

Warning - ultralytics for yolov8 is not available.
ERROR - failed to import mmyolo - MMCV==2.1.0 is used but incompatible. Please install mmcv>=2.0.0rc4, <2.1.0.

**Comment by galavera** [comment]
well, after installing and uninstalling a few times I finally got it working.
installing mmcv first using -f flag probably did the trick.
pip install mmcv==2.1.0 -f https://download.openmmlab.com/mmcv/dist/cu121/torch2.1/index.html
